### ISSUE TITLE
关于测试protocol

### ISSUE THREAD

**Issue by ConanCui** [librahu/MCRec#3]
1.你好，我看您文中的测试是说对于每一个测试物品，挑选50个负样本物品，利用他们进行排序。
![image](https://user-images.githubusercontent.com/21979236/50412664-3a466100-0844-11e9-8679-245c22efed0e.png)

您的程序里是（用的movielens100k为例），对每一个测试物品时，测试其在所有物品中的排序，这跟论文中描述的不太相同，应该以哪种为主，您论文中的结果是在哪种设置下的结果？我跑了跑您的代码（在movielens上），结果看起来应该是在所有物品中的排序。

另外想请问，能否提供预处理数据得到预训练embedding （ml-100k.bpr.item_embedding，ml-100k.bpr.type_embedding，ml-100k.bpr.user_embedding）的代码呢？

**Comment by librahu** [comment]
测试部分，对于大数据集我们按照采样的方法进行排序（因为大数据集对所有样本进行排序的话非常耗时）。而对于小数据集，没有大数据集的耗时的问题，所以我们直接就对所有商品进行排序，而且这样更加准确。
预训练的话，你可以直接用metapath2vec或者hin2vec进行生成embedding。或者借助于fm或者svdfeature这些推荐框架来生成各类节点的embedding。

**Comment by sshzhang** [comment]
@librahu  hello author, can you kindly tell me how to use metapath2vec  to get the embedding vector ?  I could 't find the code. Thank you very much!

**Comment by librahu** [comment]
@sshzhang The code is available here https://ericdongyx.github.io/metapath2vec/m2v.html

**Comment by sshzhang** [comment]
@librahu  after read the meta2vec code, but I find it seem gain the node embedding  towards each one meta_path. if the node in two meta_path, the embedding will different?  is it right?   what ' more , have you try this way to make rating prediction.

**Comment by librahu** [comment]
@sshzhang  Yes， but we only need the node embeddings in one meta-path, whose performances are not bad. About the rating prediction, please refer to my TKDE paper "heterogeneous information network embedding for recommendation"

**Comment by sshzhang** [comment]
@librahu Thank for your reply soon,  I have a idea, I want to use HIN represent category  、brand ...　auxiliary information, at the same time, I want to use a Model to build reviews information, and then join together, finally make rating prediction. can you kindly give me  some suggestion. is it seem reasonable?  Thank you very much!

**Comment by librahu** [comment]
@sshzhang It is reasonable idea. The key is how to realize. One naive idea is extracting HIN feature via metapath2vec and reviews feature via LDA, respectively. But this way seems lack novelty.

**Comment by sshzhang** [comment]
yes, This is my problem . I need a proper method  extract feature from reviews. Thank you very much!

**Comment by sshzhang** [comment]
@librahu sorry bother you again , in your experiment, can you tell me which meta_path that you use to get embedding vector?  I  need to find the best embedding vector. Thank you very much

**Comment by sshzhang** [comment]
@librahu sometime I think , I could n't get  all node embedding. Can you share your preprocessing code with us ! I need your code to get the all node embedding. Thank you very much!

**Comment by librahu** [comment]
@sshzhang Since I'm not in my university now, I cannot share my original code with you. If you cannot obtain all the nodes via metapath2vec. You can use HIN2vec (https://github.com/csiesheep/hin2vec) or the embedding generating method in my TKDE paper (https://github.com/librahu/HERec) to get all node embedding you need. By the way, metapath2vec is also able to get all node embeddings, since you can feed the random walk into metapath2vec or select multiple metapath for generating node sequence.

**Comment by sshzhang** [comment]
Thank you for your kindly reply.
